### PR TITLE
[ios][platform_view] fix platform view clipping path intersection

### PR DESCRIPTION
**Pull request by hellohuanlin**
It turns out that when there are multiple paths to clip, they are unioned together, rather than intersected. But clipping paths need to be intersected. 

There isn't any good way to intersect arbitrary paths. However, it is easy to intersect rect paths, which is the most common use case. Then we simply fallback to software rendering if we have to intersect non-rect paths. That is: 

**Case 1** Only 1 clipping path (either rect path or arbitrary path):
Hardware rendering. This should be the most common use case

**Case 2** Multiple rect clipping path: 
Hardware rendering. This is also common, and it's the linked issue that we are fixing. 

**Case 3** Other complex case (multiple non-rect clipping path)
Fallback to software rendering. This should be rare. 

After https://github.com/flutter/engine/pull/53826, we don't have a working benchmark that measures the main thread anymore. However, this PR shouldn't impact our ad benchmark, since it only has 1 clipping path. I will verify manually by checking Instruments and make sure no software rendering is happening. But we really should make the benchmark working again, not just for performance improvement, but also for monitoring regression. 

*List which issues are fixed by this PR. You must list at least one issue.*

Fixes https://github.com/flutter/flutter/issues/153904

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by jonahwilliams — COMMENTED** [review]
You could also create multiple CAShape layers instead

**Comment by hellohuanlin** [comment]
It's pretty hard to review the scenario tests, so I made a video. 

The existing test only clips a single path. So I added a "multi paths" version for each, and made sure the resulting clip is the intersection of paths (rather than the union). 

https://github.com/user-attachments/assets/3dc220be-9866-44c4-aff8-8981a6dc53fb

**Comment by hellohuanlin** [comment]
> You could also create multiple CAShape layers instead

After some research, it seems the only possible way is to have nested parents, each taking one shape layer as the mask. I don't think it's possible to just have one single mask that is the intersection of multiple shape layers. So this will require much larger change, and make the already complex view hierarchy more complex.

**Comment by hellohuanlin** [comment]
Verified the scrolling ad banner example still runs smoothly on device, and avoids software rendering related instructions (by looking at the Instruments).

**Comment by hellohuanlin** [comment]
Verified the linked issue worked after this fix

Screenshot 2024-08-30 at 2 15 38 PM

**Comment by nateshmbhat** [comment]
Eagerly waiting for this PR to be merged. Our video player ui broke after we upgraded to 3.24. 🙏🏻

File: testing/scenario_app/bin/run_ios_tests.dart [hunk 1]
```diff
@@ -358,14 +358,23 @@ final _skipTestsForImpeller = [
   'ScenariosUITests/MultiplePlatformViewsTest/testPlatformView',
   'ScenariosUITests/NonFullScreenFlutterViewPlatformViewUITests/testPlatformView',
   'ScenariosUITests/PlatformViewMutationClipPathTests/testPlatformView',
+  'ScenariosUITests/PlatformViewMutationClipPathMultipleClipsTests/testPlatformView',
```
Comment (jmagman) [inline testing/scenario_app/bin/run_ios_tests.dart:361]: Oh I didn't realize these were skipped for impeller... https://github.com/flutter/flutter/issues/131888#issuecomment-2327112432

**Review by jonahwilliams — COMMENTED** [review]
Approach makes sense. Its unfortunate to lose the hardware acceleration, but in most cases I suspect we'll only ever have rect clips.

Skia paths do have some ability to compute intersections/overlaps via skia path ops, which we definitely have access to: https://api.skia.org/SkPathOps_8h.html

Have you looked at using these methods to compute the path intersections?

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by kishangohel** [comment]
@hellohuanlin can we cherry-picked this to the current stable channel? As it's most required fixed ASAP

**Comment by hellohuanlin** [comment]
Yes we should cherry pick this. Thanks for creating the request.

**Pull request merged**